pkg/eventqueue/eventqueue.go

```diff
@@ -0,0 +1,276 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eventqueue
+
+import (
+	"reflect"
+	"sync"
+
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/spanstat"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "eventqueue")
+)
+
+// EventQueue is a structure which is utilized to handle Events in a first-in,
+// first-out order. An EventQueue may be closed, in which case all events which
+// are queued up, but have not been processed yet, will be cancelled (i.e., not
+// ran). It is guaranteed that no events will be scheduled onto an EventQueue
+// after it has been closed; if any event is attempted to be scheduled onto an
+// EventQueue after it has been closed, it will be cancelled immediately. For
+// any event to be processed by the EventQueue, it must implement the
+// `EventHandler` interface. This allows for different types of events to be
+// processed by anything which chooses to utilize an `EventQueue`.
+type EventQueue struct {
+	// This should always be a buffered channel.
+	events chan *Event
+
+	// close is closed once the EventQueue has been closed.
+	close chan struct{}
+
+	// drain is closed when the EventQueue is stopped. Any Event which is
+	// Enqueued after this channel is closed will be cancelled / not processed
+	// by the queue. If an Event has been Enqueued, but has not been processed
+	// before this channel is closed, it will be cancelled and not processed
+	// as well.
+	drain chan struct{}
+
+	// eventQueueOnce is used to ensure that the EventQueue business logic can
+	// only be ran once.
+	eventQueueOnce sync.Once
+
+	// closeOnce is used to ensure that the EventQueue can only be closed once.
+	closeOnce sync.Once
+
+	// closeWaitGroup ensures that the events channel is not closed before all
+	// events have been consumed off of it.
+	closeWaitGroup sync.WaitGroup
+}
+
+// NewEventQueue returns an EventQueue with a capacity for only one event at
+// a time, and all other needed fields initialized.
+func NewEventQueue() *EventQueue {
+	return &EventQueue{
```

Comment:
I would return `NewEventQueueBuffered(1))`

Reply:
Will be fixed in next upload.